### PR TITLE
Added endpoints for availability restriction requests

### PR DESCRIPTION
**Pull request by AlexeyShok**
Added AvailabilityRestriction model

### CODE REVIEW

File: HappyTravel.Komoro.Api/Services/Availabilities/AvailabilityRestrictionService.cs [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+                    && ar.EndDate == restriction.EndDate 
+                    && ar.RoomType.Code == restriction.RoomTypeCode
+                    && ar.RatePlanCode == restriction.RatePlanCode)
+                .ToListAsync();
+            var utcNow = _dateTimeOffsetProvider.UtcNow();
+
+            if (restriction.RestrictionStatus is not null)
+            {
+                var existingRestriction = existingRestrictions
+                    .SingleOrDefault(r => r.RestrictionStatus.Restriction == restriction.RestrictionStatus.Restriction);
```
Comment (hasmikharutyunyan) [inline HappyTravel.Komoro.Api/Services/Availabilities/AvailabilityRestrictionService.cs:53]: For me a little bit not understandable that restriction have status which inside have property Restriction.

File: HappyTravel.Komoro.Api/Services/Availabilities/AvailabilityRestrictionService.cs [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+                    };
+                    _komoroContext.AvailabilityRestrictions.Add(newRestriction);
+                }
+                else
+                {
+                    existingRestriction.RestrictionStatus = restriction.RestrictionStatus;
+                    existingRestriction.Modified = utcNow;
+                    _komoroContext.AvailabilityRestrictions.Update(existingRestriction);
+                }
+            }
```
Comment (hasmikharutyunyan) [inline HappyTravel.Komoro.Api/Services/Availabilities/AvailabilityRestrictionService.cs:107]: Could this logic will more understandable if logic inside `if else` blocks will be separated into some local functions ?

File: HappyTravel.Komoro.Data/KomoroContext.cs [hunk 3]
```diff
@@ -99,6 +101,23 @@ protected override void OnModelCreating(ModelBuilder builder)
...
+        {
+            e.ToTable("AvailabilityRestrictions");
+            e.HasKey(ar => ar.Id);
+            e.HasIndex(ar => ar.PropertyId);
+            e.Property(ar => ar.StartDate).IsRequired();
+            e.Property(ar => ar.EndDate).IsRequired();
+            e.HasIndex(ar => ar.RoomTypeId);
+            e.Property(ar => ar.RatePlanCode).IsRequired();
+            e.Property(ar => ar.RestrictionStatus).HasColumnType("jsonb");
+            e.Property(ar => ar.LengthOfStay).HasColumnType("jsonb");
```
Comment (hasmikharutyunyan) [inline HappyTravel.Komoro.Data/KomoroContext.cs:115]: For me LengthOfStay is associated with number, so maybe change the property name?

File: HappyTravel.Komoro.TravelClickChannelManager/Infrastructure/Extensions/AvailStatusMessageExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+                "Close" => RestrictionStatuses.Close,
+                _ => throw new NotImplementedException(),
+            };
+        }
+
+
+        static int GetMinAdvancedBookingOffset(string minAdvancedBookingOffset)
+        {
+            var numberOfDays = 0;
+            if (minAdvancedBookingOffset.Length >= 3)
```
Comment (hasmikharutyunyan) [inline HappyTravel.Komoro.TravelClickChannelManager/Infrastructure/Extensions/AvailStatusMessageExtensions.cs:68]: maybe needed this values to store in constant ?

File: HappyTravel.Komoro.TravelClickChannelManager/Infrastructure/Extensions/AvailabilityRestrictionExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,68 @@
+﻿using HappyTravel.Komoro.TravelClickChannelManager.Models.Availabilities;
+using HappyTravel.KomoroContracts.Availabilities;
+
+namespace HappyTravel.Komoro.TravelClickChannelManager.Infrastructure.Extensions;
+
+public static class AvailabilityRestrictionExtensions
+{
+    public static AvailStatusMessage ToAvailStatusMessage(this AvailabilityRestriction availabilityRestriction)
```
Comment (hasmikharutyunyan) [inline HappyTravel.Komoro.TravelClickChannelManager/Infrastructure/Extensions/AvailabilityRestrictionExtensions.cs:8]: I think it is better to use Full names

File: HappyTravel.Komoro.TravelClickChannelManager/Infrastructure/Extensions/AvailabilityRestrictionExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    {
+        return new AvailStatusMessage
+        {
+            StatusApplicationControl = new StatusApplicationControl
+            {
+                Start = availabilityRestriction.StartDate.ToDateTime(TimeOnly.MinValue),
+                End = availabilityRestriction.EndDate.ToDateTime(TimeOnly.MinValue),
+                InvTypeCode = availabilityRestriction.RoomTypeCode,
```
Comment (hasmikharutyunyan) [inline HappyTravel.Komoro.TravelClickChannelManager/Infrastructure/Extensions/AvailabilityRestrictionExtensions.cs:16]: Here also, I think better to use full name

File: HappyTravel.Komoro.TravelClickChannelManager/Infrastructure/Extensions/AvailabilityRestrictionExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        static LengthsOfStay GetLengthsOfStay(KomoroContracts.Availabilities.LengthOfStay lengthsOfStay)
+        {
+            return new LengthsOfStay
+            {
+                ArrivalDateBased = lengthsOfStay.IsArrivalDateBased,
+                LengthOfStayList = new List<Models.Availabilities.LengthOfStay>
+                {
+                    new Models.Availabilities.LengthOfStay
+                    {
+                        MinMaxMessageType = "MinLOS",
```
Comment (hasmikharutyunyan) [inline HappyTravel.Komoro.TravelClickChannelManager/Infrastructure/Extensions/AvailabilityRestrictionExtensions.cs:54]: maybe store these values in constants ?

File: HappyTravel.Komoro.TravelClickChannelManager/Infrastructure/Extensions/AvailStatusMessageExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+                    : null
+            };
+        }
+
+
+        static RestrictionTypes GetRestriction(string restriction)
+        {
+            return restriction switch
+            {
+                "Master" => RestrictionTypes.Master,
```
Comment (hasmikharutyunyan) [inline HappyTravel.Komoro.TravelClickChannelManager/Infrastructure/Extensions/AvailStatusMessageExtensions.cs:47]: maybe create constants for these values ?

File: HappyTravel.Komoro.Data/KomoroContext.cs [hunk 9]
```diff
@@ -99,6 +101,23 @@ protected override void OnModelCreating(ModelBuilder builder)
...
+        builder.Entity<AvailabilityRestriction>(e =>
+        {
+            e.ToTable("AvailabilityRestrictions");
+            e.HasKey(ar => ar.Id);
+            e.HasIndex(ar => ar.PropertyId);
+            e.Property(ar => ar.StartDate).IsRequired();
+            e.Property(ar => ar.EndDate).IsRequired();
+            e.HasIndex(ar => ar.RoomTypeId);
+            e.Property(ar => ar.RatePlanCode).IsRequired();
+            e.Property(ar => ar.RestrictionStatus).HasColumnType("jsonb");
```
Comment (hasmikharutyunyan) [inline HappyTravel.Komoro.Data/KomoroContext.cs:114]: I think `RestrictionStatus` could be renamed to more descriptive name

File: HappyTravel.Komoro.Api/Services/Availabilities/AvailabilityRestrictionService.cs (hunk 1 above)
Comment (AlexeyShok) [reply HappyTravel.Komoro.Api/Services/Availabilities/AvailabilityRestrictionService.cs:53]: These names are taken from the supplier's models. Can you suggest a better option?

File: HappyTravel.Komoro.Api/Services/Availabilities/AvailabilityRestrictionService.cs (hunk 2 above)
Comment (AlexeyShok) [reply HappyTravel.Komoro.Api/Services/Availabilities/AvailabilityRestrictionService.cs]: Fixed

File: HappyTravel.Komoro.TravelClickChannelManager/Infrastructure/Extensions/AvailabilityRestrictionExtensions.cs (hunk 6 above)
Comment (AlexeyShok) [reply HappyTravel.Komoro.TravelClickChannelManager/Infrastructure/Extensions/AvailabilityRestrictionExtensions.cs:16]: This is the supplier model

File: HappyTravel.Komoro.TravelClickChannelManager/Infrastructure/Extensions/AvailStatusMessageExtensions.cs (hunk 8 above)
Comment (AlexeyShok) [reply HappyTravel.Komoro.TravelClickChannelManager/Infrastructure/Extensions/AvailStatusMessageExtensions.cs]: Fixed

File: HappyTravel.Komoro.TravelClickChannelManager/Infrastructure/Extensions/AvailabilityRestrictionExtensions.cs (hunk 7 above)
Comment (AlexeyShok) [reply HappyTravel.Komoro.TravelClickChannelManager/Infrastructure/Extensions/AvailabilityRestrictionExtensions.cs]: Fixed

File: HappyTravel.Komoro.TravelClickChannelManager/Infrastructure/Extensions/AvailStatusMessageExtensions.cs (hunk 4 above)
Comment (AlexeyShok) [reply HappyTravel.Komoro.TravelClickChannelManager/Infrastructure/Extensions/AvailStatusMessageExtensions.cs:68]: I don't think it's necessary

File: HappyTravel.Komoro.TravelClickChannelManager/Infrastructure/Extensions/AvailabilityRestrictionExtensions.cs (hunk 5 above)
Comment (AlexeyShok) [reply HappyTravel.Komoro.TravelClickChannelManager/Infrastructure/Extensions/AvailabilityRestrictionExtensions.cs:8]: This is the supplier model

File: HappyTravel.Komoro.Data/KomoroContext.cs (hunk 3 above)
Comment (AlexeyShok) [reply HappyTravel.Komoro.Data/KomoroContext.cs:115]: This name includes not only the number of days, but also from what moment these days are counted.

File: HappyTravel.Komoro.Data/KomoroContext.cs (hunk 9 above)
Comment (AlexeyShok) [reply HappyTravel.Komoro.Data/KomoroContext.cs:114]: These names are taken from the supplier's models. Can you suggest a better option?

Comment (hasmikharutyunyan) [reply HappyTravel.Komoro.Data/KomoroContext.cs:114]: maybe use attributes for supplier names and name Models to more understandable ?

File: HappyTravel.Komoro.Data/KomoroContext.cs (hunk 3 above)
Comment (hasmikharutyunyan) [reply HappyTravel.Komoro.Data/KomoroContext.cs:115]: maybe somethink like `StayDurationInfo` ?

File: HappyTravel.Komoro.Api/Services/Availabilities/AvailabilityRestrictionService.cs (hunk 1 above)
Comment (hasmikharutyunyan) [reply HappyTravel.Komoro.Api/Services/Availabilities/AvailabilityRestrictionService.cs]: `RestrictionInfo` or `RestrictionDetails` ?

**Review by oleg-n-k — APPROVED** [review]

**Review by hasmikharutyunyan — APPROVED** [review]

File: HappyTravel.Komoro.Data/KomoroContext.cs (hunk 3 above)
Comment (AlexeyShok) [reply HappyTravel.Komoro.Data/KomoroContext.cs]: Renamed to StayDurationDetails

File: HappyTravel.Komoro.Data/KomoroContext.cs (hunk 9 above)
Comment (AlexeyShok) [reply HappyTravel.Komoro.Data/KomoroContext.cs]: Renamed to RestrictionStatusDetails

**Pull request merged**